pkg/traceql/engine.go

```diff
@@ -99,6 +102,114 @@ func (e *Engine) Execute(ctx context.Context, searchReq *tempopb.SearchRequest,
 	return res, nil
 }
 
+func (e *Engine) ExecuteTagValues(
+	ctx context.Context,
+	req *tempopb.SearchTagValuesRequest,
+	cb func(v Static) bool,
+	fetcher SpansetFetcher,
+) error {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "traceql.Engine.ExecuteTagValues")
+	defer span.Finish()
+
+	query := extractMatchers(req.Query)
+
+	span.SetTag("query", req.Query)
+	span.SetTag("sanitized query", query)
+
+	rootExpr, err := Parse(query)
+	if err != nil {
+		return err
+	}
+	if err := rootExpr.validate(); err != nil {
+		return err
+	}
+
+	searchReq := &tempopb.SearchRequest{
+		Start: 0, // TODO: Should add Start and End
+		End:   math.MaxUint32,
+	}
+
+	attribute, err := ParseIdentifier(req.TagName)
+	if err != nil {
+		return err
+	}
+
+	fetchSpansRequest := e.createFetchSpansRequest(searchReq, rootExpr.Pipeline)
+	// TODO: remove other conditions for the attribute we're searching for
+	// for _, cond := range fetchSpansRequest.Conditions {
+	// 	if cond.Attribute == attribute {
+	// 		return fmt.Errorf("cannot search for tag values for tag that is already used in query")
+	// 	}
+	// }
+	fetchSpansRequest.Conditions = append(fetchSpansRequest.Conditions, Condition{
+		Attribute: attribute,
+		Op:        OpNone,
+	})
+
+	span.SetTag("pipeline", rootExpr.Pipeline)
+	span.SetTag("fetchSpansRequest", fetchSpansRequest)
+
+	// set up the expression evaluation as a filter to reduce data pulled
+	fetchSpansRequest.Filter = func(inSS *Spanset) ([]*Spanset, error) {
+		if len(inSS.Spans) == 0 {
+			return nil, nil
+		}
+
+		evalSS, err := rootExpr.Pipeline.evaluate([]*Spanset{inSS})
+		if err != nil {
+			span.LogKV("msg", "pipeline.evaluate", "err", err)
+			return nil, err
+		}
+
+		if len(evalSS) == 0 {
+			return nil, nil
+		}
+
+		for _, ss := range evalSS {
+			for _, s := range ss.Spans {
+				switch attribute.Scope {
+				case AttributeScopeNone: // If tag is unscoped, we need to check all tags by name
+					for attr, v := range s.Attributes() {
+						if attr.Name == attribute.Name {
+							// TODO: Should we just quit if cb returns true?
+							cb(v) // Can't exit early because resource and span scoped tags may have the same name
```

Comment:
why can't we exit early? `cb()` returning false just indicates that the distinct value collector has crossed the limit right? it has nothing to do with span vs resource vs no scope

Reply:
You're right. I misunderstood that return.